File: csharp/cars-assemble/CarsAssemble.cs
static class AssemblyLine
{
    private const int BaseProductionRatePerHour = 221;
    private const int MinutesInHour = 60;

    public static double SuccessRate(int speed) =>
        speed switch
        {
            >= 1 and <= 4 => 1.0,
            >= 5 and <= 8 => 0.9,
            9 => 0.8,
            10 => 0.77,
            _ => 0.0
        };

    public static double ProductionRatePerHour(int speed) => speed * BaseProductionRatePerHour * SuccessRate(speed);
    public static int WorkingItemsPerMinute(int speed) => (int)(ProductionRatePerHour(speed) / MinutesInHour);
}

File: fsharp/cars-assemble/CarsAssemble.fs
module CarsAssemble

let (|InRange|_|) low high x =
  if low <= x && x <= high then Some () else None

let successRate (speed: int): float =
    match speed with
    | 0 -> 0.0
    | InRange 1 4 -> 1.0
    | InRange 5 8 -> 0.9
    | 9 -> 0.8
    | 10 -> 0.77
    | _ -> failwith $"Speed of {speed} is out of the question"

let productionRatePerHour (speed: int): float =
    let baseProductionPerHour = 221
    float (speed * baseProductionPerHour) * successRate(speed)

let workingItemsPerMinute (speed: int): int =
    int (productionRatePerHour speed / 60.0)
